migrations/versions/0379_update_archived_users.py

```diff
@@ -0,0 +1,112 @@
+    #   If they have an archive_user event in the events table, get the event_id
+    #   Get their current user email address (_archived_ format)
+    #   Get the date they were archived by extracting it from their archived email address
+    #   Get their original email address by extracting it from the un-redacted archived email address.
+    results = conn.execute(
+        text(
+            textwrap.dedent(
+                """
+                SELECT
+                    users.id as user_id,
+                    events.id AS event_id,
+                    substring(users.email_address, '_archived_(\d{4}-\d{2}-\d{2})_.*') AS archival_date,
+                    substring(users.email_address, '_archived_\d{4}-\d{2}-\d{2}_(.*)') AS original_email_address
+                FROM users
+                LEFT JOIN events on users.id = (events.data->>'user_id')::uuid AND events.event_type = 'archive_user'
+                WHERE
+                    users.email_address ~ '_archived_\d{4}-\d{2}-\d{2}_(.*)'
+                    AND NOT users.email_address LIKE CONCAT('_archived_%@', :notify_email_domain)
+                ORDER BY
+                    users.id;
+                """  # noqa: W605
+            )
+        ),
+        notify_email_domain=current_app.config["NOTIFY_EMAIL_DOMAIN"],
+    )
+
+    rows = results.fetchall()
+
+    email_domain = current_app.config["NOTIFY_EMAIL_DOMAIN"]
+    users_to_update = [
+        {
+            "user_id": user_id,
+            "desired_email_address": get_archived_db_column_value(f"{user_id}@{email_domain}", date=archival_date),
+        }
+        for user_id, _, archival_date, _ in rows
+    ]
+    print(f"Updating {len(users_to_update)} users.")
+    if users_to_update:
+        conn.execute(
+            text(
+                textwrap.dedent(
+                    """
+                    UPDATE users
+                    SET
+                        name = 'Archived user',
+                        email_address = :desired_email_address,
+                        updated_at = now()
+                    WHERE
+                        id = :user_id
+                    """
+                )
+            ),
+            users_to_update,
+        )
+
+    events_to_update = [
+        {
+            "event_id": event_id,
+            "original_email_address": original_email_address,
+        }
+        for _, event_id, _, original_email_address in rows
+        if event_id is not None
+    ]
+    print(f"Updating {len(events_to_update)} events.")
+    if events_to_update:
+        conn.execute(
+            text(
+                textwrap.dedent(
+                    """
+                    UPDATE events
+                    SET
+                        data = events.data::jsonb || jsonb_build_object('user_email_address', :original_email_address)
+                    WHERE
+                        id = :event_id
+                        AND events.data->>'user_email_address' IS NULL
```

Comment:
minor edge case, but user `fcce3265-4a2b-4763-a4a3-d9a6b0d0034c` (an address associated with chris's personal account) has four events in the events table from last november that arent captured by this query cos the user has been unarchived (email address is normal, state is active). But the old archive events are still there without an email address.

i wonder if the two queries should run from separate data sources (and in the other order to ensure that we get the orig email address.

so in sequelly pseudocode (but we might still need the python step to do some data manip)

1. `update events set user_email_address = users.email_address from users where events.data->>'user_id' = users.id and user_email_address is null`
2. `update users set email_address = get_archived_db_column_value(...) where users.state = 'inactive'`

Reply:
I did spot this and decided we didn't care, because it's just one of our dev accounts and will drop out of the events table soon anyway...

Reply:
👍